runtime/ints.c

```diff
@@ -25,6 +25,14 @@
 #include "caml/misc.h"
 #include "caml/mlvalues.h"
 
+/* Comparison resulting in -1,0,1, without extra integer width conversion */
+#define COMPARE_INT(type, v1, v2, result) \
+  type __compare_v1 = v1; \
+  type __compare_v2 = v2; \
+  intnat __compare_greater = __compare_v1 > __compare_v2; \
+  intnat __compare_less    = __compare_v1 < __compare_v2; \
+  intnat result = __compare_greater - __compare_less
+
```

Comment:
I'm not a fan of macros that declare variables in the current scope.  (As opposed to within a block.)  I'm not sure macros are needed anyway, see below.

Reply:
A macro is nice because there's one canonical reference for the code, not a couple of places where it's inlined. That helps with maintenance -- only one place to change, and it's also discoverable by those new to the codebase.
And we also don't have to document the reasoning for the code in more than one place.

I'd be happy to change the style of the macro, but I'm not exactly sure how.
One way could be make the entire function the macro, and have something like
`DECLARE_COMPARE_UNBOXED(caml_int32_compare_unboxed,int32_t)`
Another would be to take the one-liner you propose below, and make that into a macro.

Reply:
The suggestion, if I understand it correctly, was to do something akin to:

```
#define COMPARE_INT(type, v1, v2, result) \
  { type __compare_v1 = v1; \
    type __compare_v2 = v2; \
    intnat __compare_greater = __compare_v1 > __compare_v2; \
    intnat __compare_less    = __compare_v1 < __compare_v2; \
    intnat result = __compare_greater - __compare_less }
```
following the style of *e.g.* https://github.com/ocaml/ocaml/blob/trunk/runtime/interp.c#L70.

I think it is a bit safer/better than your change in https://github.com/ocaml/ocaml/pull/2250/commits/1df410b987fdb3b5dedbc2c3083a13b656b1c144, because
we are sure the expressions passed to the macro is not be evaluated
mutilple times.